### PR TITLE
PCH: Enhance Request Handling in Provider Classes

### LINKED ISSUE #1928: Use AbortController in our data fetching code
## Is your feature request related to a problem?
We currently aren't using [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) when fetching data in our React code.

## Describe the solution you'd like
See if/how we can implement `AbortController` in our React code that's fetching data.

### PR DESCRIPTION
**Pull request by vaurdan**
## Description
This PR refactors all provider classes to inherit from a newly introduced `BaseProvider`. This new base class equips each provider with an enhanced method for fetching data from the API that includes support for cancelling ongoing requests using the `AbortController`.

This update transitions provider methods from static to instance methods, which allows each provider to maintain its own map of `AbortController` instances, one per request. This architecture ensures that cancelling one request does not inadvertently cancel others. While not currently implemented, this setup facilitates future enhancements, such as assigning unique abort IDs for selectively cancelling requests in complex operations like smart linking across multiple paragraphs.

This architecture not only simplifies the management of request cancellations but also positions the codebase for more sophisticated request handling scenarios in the future.

## Motivation and Context
- Implements cancellable requests, thereby resolving issue #1928.
- Enhances the reliability of the codebase and reduces redundancy by utilizing inheritance for common provider functionalities.

## How has this been tested?
Each provider was tested individually through manual checks, end-to-end tests, and unit tests with Jest.

## Summary by CodeRabbit

- **New Features**
	- Introduced a base class for improved data fetching with request cancellation capabilities.
	- Enhanced smart linking and title suggestion functionalities by utilizing a singleton pattern for better resource management.

- **Bug Fixes**
	- Removed the `isDismissible` property from certain UI components to streamline user notifications.

- **Documentation**
	- Standardized JSDoc comments across various components for better consistency and readability.

- **Refactor**
	- Updated several components to use singleton access methods, aligning with modern best practices.

- **Chores**
	- Updated dependencies to improve stability and performance of testing frameworks.

### CODE REVIEW

**Comment by vaurdan** [comment]
@coderabbitai review

File: src/content-helper/common/base-provider.tsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ * The response structure of the API.
+ *
+ * @since 3.15.0
+ */
+export interface ContentHelperAPIResponse<T> {
+	error?: Error;
+	data: T;
+}
+
+type GetAbortControllerResult = {
```
Comment (acicovic) [inline src/content-helper/common/base-provider.tsx:28]: We can add a comment here if we think it would be useful.

**Review by acicovic — APPROVED** [review]
Thank you for your work and for the great PR description!

Approving with a minor comment.

File: src/content-helper/editor-sidebar/title-suggestions/provider.ts [hunk 2]
```diff
@@ -1,75 +1,75 @@
...
 	 *
+	                                  <<<<<<< HEAD
+	                                  <<<<<<< HEAD
+	 * @return { Promise<string[]> } The resulting list of titles.
+	  =======
 	 * @return {Promise<string[]>} The resulting list of titles.
+	  >>>>>>> 3696940e (TypeScript DocBlocks: Remove spaces in braces)
+	  =======
+	 * @return {Promise<string[]>} The resulting list of titles.
+	  >>>>>>> develop
```
Comment (acicovic) [inline src/content-helper/editor-sidebar/title-suggestions/provider.ts]: Some funky stuff happening here 🙂

Comment (acicovic) [reply src/content-helper/editor-sidebar/title-suggestions/provider.ts]: It also seems some additional things were reverted in https://github.com/Parsely/wp-parsely/pull/2401/commits/9396e93e48a58cb7def1b091d9638ddb15faf1bd? A bit strange.

Comment (acicovic) [reply src/content-helper/editor-sidebar/title-suggestions/provider.ts]: The revert seems to be mostly the changes of https://github.com/Parsely/wp-parsely/pull/2401/commits/e9d8c5f81ad6fa1af658792d4c305bbfc28241d2.

Comment (vaurdan) [reply src/content-helper/editor-sidebar/title-suggestions/provider.ts]: This is weird! Let me clean this up.

**Comment by vaurdan** [comment]
@acicovic the merge mess should be fixed! Not sure what happened there :D 

I took a look at the change set, and it looks alright. But just in case, feel free to take another look and let me know if I have 🟢 to merge!

**Review by acicovic — APPROVED** [review]
I think it's OK now.

**Pull request merged**